netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,420 @@
+        cancelPendingStreams();
+      }
+
+      @Override
+      public void streamInactive(Http2Stream stream) {
+        createNextPendingStream();
+      }
+    });
+  }
+
+  @Override
+  public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                                    int padding, boolean endStream, ChannelPromise promise) {
+    return writeHeaders(ctx, streamId, headers, 0, Http2CodecUtil.DEFAULT_PRIORITY_WEIGHT, false,
+                        padding, endStream, promise);
+  }
+
+  @Override
+  public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                                    int streamDependency, short weight, boolean exclusive,
+                                    int padding, boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId) || connection().isGoAway()) {
+      return dispatchEncoder.writeHeaders(ctx, streamId, headers, streamDependency, weight,
+                                          exclusive, padding, endOfStream, promise);
+    }
+    if (connection().local().acceptingNewStreams()) {
+      assert streamId > largestCreatedStreamId;
+      largestCreatedStreamId = streamId;
+      return dispatchEncoder.writeHeaders(ctx, streamId, headers, streamDependency, weight,
+              exclusive, padding, endOfStream, promise);
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream == null) {
+      pendingStream = new PendingStream(ctx, streamId);
+      pendingStreams.put(streamId, pendingStream);
+    }
+    pendingStream.frames.add(new HeadersFrame(headers, streamDependency, weight, exclusive,
+                             padding, endOfStream, promise));
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
+                                      ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return dispatchEncoder.writeRstStream(ctx, streamId, errorCode, promise);
+    }
+    /**
+     * Since the {@code dispatchEncoder} doesn't know about any buffered streams
+     * we have to handle cancellation of the promises and releasing of the ByteBufs
+     * here.
+     */
+    PendingStream stream = pendingStreams.remove(streamId);
+    if (stream != null) {
+      cancelPendingStream(stream);
+      promise.setSuccess();
+    } else {
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                 boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return dispatchEncoder.writeData(ctx, streamId, data, padding, endOfStream, promise);
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
```

Comment:
@nmittler I kinda wanted to avoid this in order to be fast. With the current implementation the fast-path is a integer comparison `streamId <= largestCreatedStreamId`. With your code we would have to do a table lookup for every single data frame.